File: HeroesVillainsExample/HeroesVillainsExample.py
import sys

def main():
    run_again = True
    while run_again:
        print("\nWelcome to the Heroes vs Villains Battle Simulator!")
        print("You will need to enter the following values:")
        print("1. Number of Villains (N)")
        print("2. Number of Heroes (M)")
        print("3. Health of each Hero (H)")        
        print("4. Health of each Villain (one per line, total of N values)")
        print("--------------------------------------------------------")

        try:

            # 1. Read Number of Villains (N)
            while True:
                try:
                    N = int(input("Enter the number of villains (N): "))
                    # N must be a positive integer as per problem constraints (1 to 2e5)
                    # However, for the binary search, N can be 0 if all villains are skipped.
                    # We'll allow N=0 here for robustness, although problem statement implies N>=1.
                    if N < 0:
                        print("Error: Please enter a non-negative integer for N!")
                        continue
                    break
                except ValueError:
                    print("Error: Please enter a non-negative integer for N!")

            # 2. Read Number of Heroes (M)
            while True:
                try:
                    M = int(input("Enter the number of heroes (M): "))
                    # M must be a positive integer as per problem constraints (1 to 2e5)
                    if M < 1:
                        print("Error: Please enter a positive integer for M!")
                        continue
                    break
                except ValueError:
                    print("Error: Please enter a positive integer for M!")

            # 3. Read Health of each Hero (H)
            while True:
                try:
                    H = int(input("Enter the health of each hero (H): "))
                    # H must be a positive integer as per problem constraints (1 to 10^9)
                    if H < 1:
                        print("Error: Please enter a positive integer for H!")
                        continue
                    break
                except ValueError:
                    print("Error: Please enter a positive integer for H!")
            
            # 4. Read Health of each Villain (V)
            V = []
            if N > 0: # Only ask for villain health if N > 0
                print(f"Enter the health of each villain (one per line, {N} values):")
                for i in range(N):
                    while True:
                        try:
                            v = int(input(f"Villain {i+1}: "))
                            if v < 1:
                                print("Error: Please enter a positive integer for each villain's health!")
                                continue
                            V.append(v)
                            break
                        except ValueError:
                            print("Error: Please enter a positive integer for each villain's health!")

            # 5. Calculate the minimum number of villains to skip using binary search
            left = 0
            right = N # It's possible to remove all N villains
            answer = N # Initialize with the worst-case scenario (remove all villains)

            while left <= right:
                mid = (left + right) // 2 # Calculate mid using integer division
                
                # Simulate battle for villains starting from index 'mid'
                possible = True
                current_hero_count = 0 # Tracks how many heroes have been used/defeated
                current_hero_health = H # Health of the *current* hero fighting

                # Use a while loop to control villain index 'i' manually
                i = mid 
                while i < N:
                    # If all heroes are defeated before all villains
                    if current_hero_count >= M:
                        possible = False
                        break # Heroes lose, cannot defeat remaining villains

                    villain_health = V[i]

                    if current_hero_health > villain_health:
                        # Hero wins: hero's health decreases, villain is defeated.
                        current_hero_health -= villain_health
                        i += 1 # Move to the next villain
                    elif current_hero_health < villain_health:
                        # Villain wins: current hero is defeated.
                        # The *same villain* remains to fight the next hero.
                        current_hero_count += 1 # One hero used/defeated
                        current_hero_health = H # New hero steps up with full health
                        # i does NOT increment, the same villain (V[i]) will fight the next hero
                    else: # current_hero_health == villain_health (Both are defeated)
                        # Both are defeated.
                        # Next hero steps up with full health. Move to the next villain.
                        current_hero_count += 1 # One hero used/defeated
                        current_hero_health = H # New hero starts with full health
                        i += 1 # Move to the next villain
                
                # After simulating the battle for 'mid' skipped villains:
                if possible:
                    # If it's possible to win, try to skip fewer villains
                    answer = mid
                    right = mid - 1
                else:
                    # If not possible, need to skip more villains
                    left = mid + 1

            # 6. Output the result
            print(f"\nMinimum number of villains to skip from the front: {answer}")

        except Exception as ex:
            print(f"An unexpected error occurred: {ex}")
            print("Please try again.")
            continue

        # 7. Ask if user wants to run again
        user_choice = input("\nDo you want to run the program again? (Y/N): ").strip().lower()
        if user_choice not in ('y', 'yes'):
            run_again = False
            print("\nThank you for using the Heroes vs Villains Battle Simulator!")

if __name__ == "__main__":
    main()

File: HeroesVillainsExample/HeroesVillainsExample.cs
using System;

namespace HeroVillainExample
{
    public class Program
    {
        public static void Main()
        {
            bool runAgain = true;
            while (runAgain)
            {
                Console.WriteLine("\nWelcome to the Heroes vs Villains Battle Simulator!");
                Console.WriteLine("You will need to enter the following values:");
                Console.WriteLine("1. Number of Villains (N)");
                Console.WriteLine("2. Number of Heroes (M)");
                Console.WriteLine("3. Health of each Hero (H)");
                Console.WriteLine("4. Health of each Villain (one per line, total of N values)");
                Console.WriteLine("--------------------------------------------------------");

                int N = 0, M = 0, H = 0;
                int[] V = null;

                try
                {
                    // 1. Read Number of Villains (N)
                    Console.Write("Enter the number of villains (N): ");
                    if (!int.TryParse(Console.ReadLine(), out N) || N < 1)
                    {
                        Console.WriteLine("Error: Please enter a positive integer for N!");
                        continue;
                    }

                    // 2. Read Number of Heroes (M)
                    Console.Write("Enter the number of heroes (M): ");
                    if (!int.TryParse(Console.ReadLine(), out M) || M < 1)
                    {
                        Console.WriteLine("Error: Please enter a positive integer for M!");
                        continue;
                    }

                    // 3. Read Health of each Hero (H)
                    Console.Write("Enter the health of each hero (H): ");
                    if (!int.TryParse(Console.ReadLine(), out H) || H < 1)
                    {
                        Console.WriteLine("Error: Please enter a positive integer for H!");
                        continue;
                    }

                    // 4. Read Health of each Villain (V)
                    Console.WriteLine($"Enter the health of each villain (one per line, {N} values):");
                    V = new int[N];
                    for (int i = 0; i < N; i++)
                    {
                        Console.Write($"Villain {i + 1}: ");
                        if (!int.TryParse(Console.ReadLine(), out V[i]) || V[i] < 1)
                        {
                            Console.WriteLine("Error: Please enter a positive integer for each villain's health!");
                            i--; // Try this villain again
                            continue;
                        }
                    }

                    // 5. Calculate the minimum number of villains to skip
                    int left = 0, right = N, answer = N;
                    while (left <= right)
                    {
                        int mid = (left + right) / 2;
                        bool possible = true;
                        int heroIndex = 0;
                        long currentHeroHealth = H;
                        for (int i = mid; i < N; i++)
                        {
                            if (V[i] > H)
                            {
                                possible = false;
                                break;
                            }
                            if (currentHeroHealth > V[i])
                            {
                                currentHeroHealth -= V[i];
                            }
                            else
                            {
                                heroIndex++;
                                if (heroIndex >= M)
                                {
                                    possible = false;
                                    break;
                                }
                                currentHeroHealth = H - V[i];
                            }
                        }
                        if (possible)
                        {
                            answer = mid;
                            right = mid - 1;
                        }
                        else
                        {
                            left = mid + 1;
                        }
                    }

                    // 6. Output the result
                    Console.WriteLine($"\nMinimum number of villains to skip from the front: {answer}");

                }
                catch (Exception ex)
                {
                    Console.WriteLine($"An error occurred: {ex.Message}");
                    Console.WriteLine("Please try again.");
                    continue;
                }

                // 7. Ask if user wants to run again
                Console.Write("\nDo you want to run the program again? (yes/no): ");
                string userChoice = Console.ReadLine().Trim().ToLower();
                if (userChoice != "yes")
                {
                    runAgain = false;
                    Console.WriteLine("\nThank you for using the Heroes vs Villains Battle Simulator!");
                }
            }
        }
    }
}